Miminal docker support

dmp1ce opened the pull request:
This PR allows `install.sh` to run with a `--docker-install` option which installs JoinMarket without virtualenv or sudo. By not installing with virtualenv, it makes running scripts through Docker containers more straight forward. Sudo is also unnecessary because Docker containers run as root by default.

Also included is a minimal Dockerfile which can be used as a starting point for more Docker based JoinMarket services. Documentation is also provided explaining how to use the Dockerfile.

kristapsk commented:
Concept ACK

But does not work as is on a Gentoo Linux, as it requires `--user` option to `pip` if outside virtualenv.
```
ERROR: (Gentoo) Please run pip with the --user option to avoid breaking python-exec
```
So it's hard for me to test it properly.

kristapsk commented:
And you have typo in title - miminal vs minimal.

manasgandy commented:
This PR might be useful for containerization of JoinMarket for Umbrel.

dmp1ce commented:
I couldn't say. I use this patch to build my own Dockerfile and use it in a Compose configuration I have. I've never used Umbrel.

https://github.com/dmp1ce/joinmarket-DOCKERFILE

AdamISZ commented:
@kristapsk as per above, apparently this is already in active use, is this mergeable, in some version?

kristapsk commented:
@AdamISZ Will look at this. Likely needs to be tested on different platforms as it has changes in `install.sh` which can potentially break things for non-Docker users too.

AdamISZ commented:
Agreed.

AdamISZ commented:
If it's easy to logically distinguish that using the `with_jmvenv` flag results in the identical script as before, then we can avoid checking that it still works on all environments it used to work on. I realise that's not 100% trivial because there are quite a few changes to `install.sh`, but it seems reasonable, at least.

kristapsk commented on install.sh:
```diff
@@ -382,6 +396,7 @@ Options:
 --python, -p                python version (only python3 versions are supported)
 --with-qt                   build the Qt GUI
 --without-qt                don't build the Qt GUI
+--docker-install system wide install as root for minimal Docker installs
```
Alphabetical ordering and spacing.
```suggestion
--docker-install            system wide install as root for minimal Docker installs
--python, -p                python version (only python3 versions are supported)
--with-qt                   build the Qt GUI
--without-qt                don't build the Qt GUI
```

kristapsk commented on README.md:
```diff
@@ -55,6 +55,7 @@ Alternative to this "quickstart": follow the [install guide](docs/INSTALL.md).
 ### More installation guides
 
 * [Installation on Windows](docs/INSTALL.md#installation-on-windows).
+* [Installation with Docker.](docs/INSTALL.md#docker-installation)
```
nit - for consistency with others
```suggestion
* [Installation with Docker](docs/INSTALL.md#docker-installation)
```

dmp1ce commented:
I rebased to master and I didn't get any errors from running `docker build -t joinmarket-test ./`. I am running docker on Arch Linux. I'm not sure the host OS matters.

I think I got your documentation changes in my latest commit. I couldn't figure out how to git a .patch file from the Github comments so I eyeballed it.

kristapsk commented:
Have the same error about ffi.h missing. Host OS should not matter as building happens inside Docker container. I'm currently running Docker version 20.10.9, build c2ea9bc90b under Gentoo Linux.

dmp1ce commented:
Did you get the latest upstream image? I used `docker build --no-cache`.

kristapsk commented:
Actual issue is some lines above:
```
skipping libffi-3.2.1...
skipping libsodium-1.0.18...
```
For some reason on my machine it thinks libffi and libsodium are already installed, so it ends up missing libffi on JM install (last) phase...

dmp1ce commented:
Maybe in your build cache. That could be an issue with the Dockerfile setup actually. I should ignore for already built outside of Docker.

dmp1ce commented:
@kristapsk Can you try the build again with the `.dockeringore` I added?

I reproduced the error on my machine and fixed it with the `.dockerignore`.

dmp1ce commented:
I noticed the Darwin install path was missing the checks I added for Debian install. I'm not sure it is even possible to create a Darwin image but the `--docker-install` flag should work the same as in Debian now.

kristapsk commented:
Now it works also for me. ACK, I'm ready to merge this, but you should squash commits into one.
